### PR TITLE
Refactor summarization module

### PR DESCRIPTION
**Pull request by jcc37**
* Refactor summarization module to be Angular services
* Update/move tests and dataset metas for the new summarization services
* Update ChartSummarizationComponent to take summaries from services

### CODE REVIEW

File: src/services/summarization/summarization-control.service.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        case SUMMARIZATION.TREND_WEEKLY_COMPARISON_AVERAGE:
+          return this.trendWeeklyComparisonAverageSummarizationService.summaries$(config);
+        case SUMMARIZATION.TREND_WEEKLY_COMPARISON_RATE:
+          return this.trendWeeklyComparisonRateSummarizationService.summaries$(config);
+        case SUMMARIZATION.TREND_WEEKLY_ELABORATION:
+          return this.trendWeeklyElaborationSummarizationService.summaries$(config);
+        case SUMMARIZATION.TREND_WEEKLY_PATTERN:
+          return this.trendWeeklyPatternSummarizationService.summaries$(config);
+        default:
+          return of([]);
```
Comment (Rhombus33) [inline src/services/summarization/summarization-control.service.ts:51]: nit: can also test empty case and multiple metas

File: src/services/summarization/summarization-data-source.service.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+export interface Data {
+  label: string;
+  points: Point[];
+}
+
+@Injectable({
+  providedIn: 'any',
+})
+export class SummarizationDataSourceService implements OnDestroy {
```
Comment (Rhombus33) [inline src/services/summarization/summarization-data-source.service.ts:17]: nit: tests

File: src/services/summarization/summarization.service.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+@Injectable({
+  providedIn: 'any',
+})
+export abstract class SummarizationService<PointT extends Point, PropertiesT, ConfigT extends BaseConfig> implements OnDestroy {
+  destroy$ = new Subject();
+  cachedDataProperties$ = new Map<string, Observable<PropertiesT>>();
+  cachedSummaries$ = new Map<string, Observable<SummaryGroup[]>>();
+
+  hashObject(obj: Record<string, any>): string {
+    return JSON.stringify(obj, Object.keys(obj).sort());
```
Comment (Rhombus33) [inline src/services/summarization/summarization.service.ts:20]: curious, why do you need to filter and sort it?

**Review by Rhombus33 — APPROVED** [review]

File: src/services/summarization/summarization.service.ts (hunk 3 above)
Comment (jcc37) [reply src/services/summarization/summarization.service.ts:20]: so that `config = { a, b }` and `config = { b, a }` will return the same hashkey

File: src/services/summarization/summarization.service.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    return this.cachedSummaries$.get(hashKey) as Observable<SummaryGroup[]>;
+  }
+
+  properties$(rawConfig: BaseConfig & Partial<ConfigT>): Observable<PropertiesT> {
+    const config = this.prepareConfig(rawConfig);
+
+    const hashKey = this.hashObject(rawConfig);
+    if (!this.cachedProperties$.has(hashKey)) {
+      const observable = this.createProperties$(config)
```
[the remark this answers is not in the document]
Comment (jcc37) [reply src/services/summarization/summarization.service.ts]: Fixed

File: src/services/summarization/summarization.service.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      const observable = this.createSummaries$(config)
+        .pipe(shareReplay(1))
+        .pipe(takeUntil(this.destroy$));
+      this.cachedSummaries$.set(hashKey, observable);
+    }
+
+    return this.cachedSummaries$.get(hashKey) as Observable<SummaryGroup[]>;
+  }
+
+  properties$(rawConfig: BaseConfig & Partial<ConfigT>): Observable<PropertiesT> {
```
[the remark this answers is not in the document]
Comment (jcc37) [reply src/services/summarization/summarization.service.ts]: Added

File: src/services/summarization/summarization-control.service.ts (hunk 1 above)
Comment (jcc37) [reply src/services/summarization/summarization-control.service.ts:51]: added

File: src/services/summarization/summarization-data-source.service.ts (hunk 2 above)
Comment (jcc37) [reply src/services/summarization/summarization-data-source.service.ts:17]: added
